Documentation/Evolution/StringProcessingAlgorithms.md

```diff
@@ -0,0 +1,402 @@
+
+```swift
+let statement = """
+CREDIT    04/06/2020    Paypal transfer    $4.99
+CREDIT    04/03/2020    Payroll            $69.73
+DEBIT     04/02/2020    ACH transfer       ($38.25)
+DEBIT     03/24/2020    IRX tax payment    ($52,249.98)
+"""
+```
+
+We have already seen that parsing a date string can be tricky since it could contain localized month name (`"Feb"` as seen from above). Parsing a currency string such as `$3,020.85` with regex is not trivial either -- it can contain grouping separators, a decimal separator, and a currency symbol, all of which can be localized. 
+
+Foundation has various parsers for localized strings like these. Delegating this task to dedicated parsers alleviates the need to handle it yourself. In the second part of the pitch, we introduce the `CustomRegexComponent` protocol that conveniently lets types from outside the standard library participate in regex builders and `RegexProtocol` algorithms.
+
+If `Foundation.FloatingPointFormatStyle<Double>.Currency` conformed to `CustomRegexComponent`, you would be able to retrieve the currency from the bank statement above as a list of `Double` values this way:
+
+```swift
+let regex = Regex {
+    capture(.localizedCurrency(code: "USD").sign(strategy: .accounting))
+}
+
+let amount = statement.matches(of: regex).map(\.result.1)
+// [4.99, 69.73, -38.25, -52249.98]
+```
+
+
+## Detailed design
+
+### Algorithms
+
+The following regex-powered algorithms as well as their generic `Collection` equivalents are included in this pitch:
+
+#### Contains
+
+```swift
+extension Collection where Element: Equatable {
+    /// Returns a Boolean value indicating whether the collection contains the
+    /// given sequence.
+    /// - Parameter other: A sequence to search for within this collection.
+    /// - Returns: `true` if the collection contains the specified sequence,
+    /// otherwise `false`.
+    public func contains<S: Sequence>(_ other: S) -> Bool
+        where S.Element == Element
+}
+
+extension BidirectionalCollection where SubSequence == Substring {
+    /// Returns a Boolean value indicating whether the collection contains the
+    /// given regex.
+    /// - Parameter regex: A regex to search for within this collection.
+    /// - Returns: `true` if the regex was found in the collection, otherwise
+    /// `false`.
+    public func contains<R: RegexProtocol>(_ regex: R) -> Bool
+}
+```
+
+#### Starts with
+
+```swift
+extension BidirectionalCollection where SubSequence == Substring {
+    /// Returns a Boolean value indicating whether the initial elements of the
+    /// sequence are the same as the elements in the specified regex.
+    /// - Parameter regex: A regex to compare to this sequence.
+    /// - Returns: `true` if the initial elements of the sequence matches the
+    /// beginning of `regex`; otherwise, `false`.
+    public func starts<R: RegexProtocol>(with regex: R) -> Bool
+}
+```
+
+#### Trim prefix
+
+```swift
+extension Collection {
```

Comment:
I find myself squinting to see the differences between these functions. I wonder if it's easier to review if we group them by extension other than algorithms, like

```swift
extension Collection where Element: Equatable {
    public func contains<S: Sequence>(_ other: S) -> Bool
    public func trimmingPrefix<Prefix: Collection>(_ prefix: Prefix) -> SubSequence
}

extension BidirectionalCollection where SubSequence == Substring {
    public func contains<R: RegexProtocol>(_ regex: R) -> Bool
    public func starts<R: RegexProtocol>(with regex: R) -> Bool
}
```